### ISSUE TITLE
character '-' dash no longer valid environment variable

### ISSUE THREAD

**Issue by asyarif93** [compose-spec/compose-go#310]
got this error `unexpected character "-" in variable name near "AIRFLOW_CONN_COMPOSER-WORKER-SAMPLE..."`

afaik, specs for env is IEEE Std 1003.1-2001 - any printable character but = and NUL is valid

i'm not sure if the problems come from this lib or docker/compose

please close this issue if it's not relevant in this repository.

thank you

**Comment by asyarif93** [comment]
then i read the article wrong, i think we can close this. thanks for the effort

**Comment by milas** [comment]
You're both right!

Here's an expanded snippet from [IEEE Std 1003.1, 2004 Edition :: 8.1](https://pubs.opengroup.org/onlinepubs/009695399/basedefs/xbd_chap08.html) (emphasis mine):
> These strings have the form name=value; names shall not contain the character '='. For values to be portable across systems conforming to IEEE Std 1003.1-2001, the value shall be composed of **characters from the portable character set (except NUL and _as indicated below_).** There is no meaning associated with the order of strings in the environment. If more than one string in a process' environment has the same name, the consequences are undefined.
>
> Environment variable names used by the utilities in the _**Shell and Utilities**_ volume of IEEE Std 1003.1-2001 consist **solely of uppercase letters, digits, and the '_' (underscore) from the characters defined in [Portable Character Set](https://pubs.opengroup.org/onlinepubs/009695399/basedefs/xbd_chap06.html#tagtcjh_3)** and do not begin with a digit. Other characters may be permitted by an implementation; applications shall tolerate the presence of such names

Currently, Compose _MOSTLY_ behaves like a Shell - uppercase letters, digits, & `_`. We do also allow `.` as a valid character due to its prevalence in many Java frameworks for environment variable overrides.

I don't think we want to open up to all non-NUL but I could see allowing `-` if a similar argument to `.` can be made, i.e. there's widespread use of `-` from popular frameworks/languages.

---

TL;DR Stick to alphanumeric + `_` for compatibility purposes with Shell & other apps

**Comment by ndeloof** [comment]
Agree, let's reopen this one, until there's some technical reason that prevent us to accept `-` within environment variables

**Comment by xzyfer** [comment]
>TL;DR Stick to alphanumeric + _ for compatibility purposes with Shell & other apps

Looks like this rule of thumb has already been broken in https://github.com/compose-spec/compose-go/pull/271. The justification being that `.` was support in v1 which appears to be true for `-` as the current "work-around" found online is to use v1.
